File: 2021/06/22/main.py
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        r, hs, uhs = 0, set(), set()
        def is_sub(word):
            idx = -1
            for w in word:
                idx = s.find(w, idx + 1)
                if idx == -1:
                    return False
            return True
        for w in words:
            if hs.__contains__(w):
                r += 1
                continue
            if uhs.__contains__(w): continue
            if is_sub(w):
                hs.add(w)
                r += 1
            else:
                uhs.add(w)
        return r

File: 2021/06/22/main.go
package main

import "strings"

func numMatchingSubseq(s string, words []string) int {
	r, hs, uhs := 0, make(map[string]bool), make(map[string]bool)

	for _, w := range words {
		_, existh := hs[w]
		if existh {
			r++
			continue
		}
		_, existu := uhs[w]
		if existu {
			continue
		}
		if is_sub(s, w) {
			hs[w] = true
			r++
		} else {
			uhs[w] = true
		}
	}
	return r
}

func is_sub(s string, word string) bool {
	idx := 0
	for _, w := range word {
		fidx := strings.IndexRune(s[idx:], w)
		if fidx == -1 {
			return false
		}
		idx += fidx + 1
	}
	return true
}
